endpoints/openrtb2/auction.go

```diff
@@ -264,6 +265,8 @@ func (deps *endpointDeps) parseRequest(httpRequest *http.Request) (req *openrtb.
 		return
 	}
 
+	lmt.ModifyForIOS(req)
```

Comment:
Nitpick: `lmt.ModifyForIOS(req)` comes after `processInterstitials(req)` in line 263. This `processInterstitials(req)` function unmarshalls the `Device.Ext` making the `jsonparser.GetInt(deviceExt, "atts")` call that `lmt.ModifyForIOS(req)` eventually makes inside `ParseDeviceExtATTS(deviceExt json.RawMessage)` to be unnecessary. Actually, that 261 to 268 neighborhood makes use of `Device` a lot 
```
260     // Populate any "missing" OpenRTB fields with info from other sources, (e.g. HTTP request headers).
261     deps.setFieldsImplicitly(httpRequest, req)  // <--- uses `Device`
262
263     if err := processInterstitials(req); err != nil {  // <--- unmarshalls `Device.Ext` and checks for `Device!=nil` again
264         errs = []error{err}
265         return
266     }
267
268     lmt.ModifyForIOS(req)  // <--- jsonparses `Device.Ext`
269
270     errL := deps.validateRequest(req)
endpoints/openrtb2/auction.go
```
Do you think it'd be a good idea to maybe consolidate inside `deps.setFieldsImplicitly(httpRequest, req)`?
```
260       // Populate any "missing" OpenRTB fields with info from other sources, (e.g. HTTP request headers).
    +     if err := deps.setFieldsImplicitly(httpRequest, req); err != nil {
261 -     deps.setFieldsImplicitly(httpRequest, req)
262 -  
263 -     if err := processInterstitials(req); err != nil {
264           errs = []error{err}
265           return
266       }
267 -  
268 -     lmt.ModifyForIOS(req)
269
270       errL := deps.validateRequest(req)
             .
             .
             .
1059 - func (deps *endpointDeps) setFieldsImplicitly(httpReq *http.Request, bidReq *openrtb.BidRequest) {
1059 + func (deps *endpointDeps) setFieldsImplicitly(httpReq *http.Request, bidReq *openrtb.BidRequest) error {
1060 -     sanitizeRequest(bidReq, deps.privateNetworkIPValidator) // <-- READS Device and checks Device != nil
1061 -
1062 -     setDeviceImplicitly(httpReq, bidReq, deps.privateNetworkIPValidator) // <-- if Device==nil, creates one bidReq.Device = &openrtb.Device{}
     +     if bidReq.Device != nil {
     +         &openrtb.Device{}
     +     }
     +     sanitizeRequest(bidReq, deps.privateNetworkIPValidator) // <-- MODIFY ACCORDINGLY
     +     setDeviceImplicitly(httpReq, bidReq, deps.privateNetworkIPValidator) // <-- MODIFY ACCORDINGLY
     +     
     +     //unmarshal device.Ext if any     
     +     if req.Device.Ext != nil {
     +         err := json.Unmarshal(req.Device.Ext, &devExt)
     +         if err == nil {
     +             // Do the iOS and interstitials thing
     +             if err := processInterstitials(req, devExt); err != nil { // <-- MODIFY TO TAKE UNMARSHALED Device.Ext
     +                 return err // I guess we'll have to return error from setFieldsImplicitly now
     +             }
     +             lmt.ModifyForIOS(req, devExt) // <-- MODIFY TO TAKE EITHER UNMARSHALED Device.Ext OR THE ATTS VALUE
     +         }              
     +     }
1063  
1064       // Per the OpenRTB spec: A bid request must not contain both a Site and an App object.
1065       if bidReq.App == nil {
1066           setSiteImplicitly(httpReq, bidReq)
1067       }
1068       setImpsImplicitly(httpReq, bidReq.Imp)
1069  
1070       setAuctionTypeImplicitly(bidReq)
     +     return nil
1071   }
endpoints/openrtb2/auction.go
```

Reply:
On a second thought, I realize that this is more a refactoring effort than a simple nitpick and, given the need for this feature, we should probably sort it on a separate PR. 

Thoughts?

Reply:
Nevermind my comment above, sometimes I tend to overcomplicate things. A far better approach could be to simply return from `processInterstitials(req)` the unmarshaled `Device.Ext` and modify `lmt.ModifyForIOS(req)` accordingly. What do you think? Given the time constraint, do you feel it's feasible?
```
260     // Populate any "missing" OpenRTB fields with info from other sources, (e.g. HTTP request headers).
261     deps.setFieldsImplicitly(httpRequest, req)
262
263 -   if err := processInterstitials(req); err != nil {
    +   if devExt, err := processInterstitials(req); err != nil {  // MODIFY TO RETURN `devExt`
264         errs = []error{err}
265         return
266     }
267
268 -   lmt.ModifyForIOS(req)
    +   lmt.ModifyForIOS(req, devExt)  // <--- MODIFY so there's no need to jsonparse `Device.Ext` now
269
270     errL := deps.validateRequest(req)
endpoints/openrtb2/auction.go
```

Reply:
I'm unsure if this is worth fixing now, or if it's better to wait on the work @hhhjort is spearheading to solve this problem on a larger scale. Both `processInterstitials` and `ModifyForIOS` only unmarshal the device ext in certain circumstances, so I can't rely on `processInterstitials` actually performing the unmarshal.

Reply:
Agree